### ISSUE TITLE
Logging for tasks and modules

### ISSUE THREAD

**Issue by Saturate** [Saturate/polymerase#12]
We need a logging function. A log-level setting would be nice, so that we can log different things depending on a flag.

**Comment by Volzy** [comment]
Definitely. I can look at that.

**Comment by Volzy** [comment]
Where do we want to put the environment logic?

- A dump logger function that's called whenever we need to output a log message?
- A clever logger function that we always call and will determine environment before outputting log messages or not?

**Comment by Saturate** [comment]
I'm thinking we should use https://github.com/chalk/chalk and do something like

`log('INFO', 'We just did something')` If we always want to show it

`log('DEBUG', 'We just did something that might be useful for debugging to know')` If we want it with the debug flag.

`log('SILLY', 'This is just stupid, how would need all this information_!')` 

If you pick a higher log-level you will get all the lower ones as well. I don't know how many levels we will need.

INFO -> DEBUG -> SILLY

I'm thinking somebody before us did something like this, something that we can learn from or steal :-D But would be nice if you looked at it!

**Comment by Volzy** [comment]
How do you feel about this list of log levels? Seems to be the standard.

1. Off (0)
2. Fatal (100)
3. Error (200)
4. Warn (300)
5. Info (400)
6. Debug (500)
7. Trace (600)
8. All (700)

Taken from: https://logging.apache.org/log4j/2.x/manual/customloglevels.html.

If flag is set, then every log level with a lower value is outputted. E.g. --info would output info, warn, error and fatal logs.

**Comment by Saturate** [comment]
I feel awesome, seems like a very good solution :-) Nothing more to add I guess.

**Comment by Volzy** [comment]
Related PR #15

**Comment by Saturate** [comment]
Maybe we should take some inspiration from https://github.com/visionmedia/debug

**Comment by Volzy** [comment]
This has a single state (debug) and you can run debug for a limited area of your solution. How does this compare to our logging function, where we want to output log from the tasks you run across the solution based on a state you set?

Can you ellaborate on your thoughts?

**Comment by Saturate** [comment]
Yes, I guess I have to explain a bit more. As you can't readminds :(

The logger should be independent from all other components, so a logger would have to:

- Read the configuration from a **Environment Variable** this would make more sense that than a argument or configuration, as we can start logging from the very beginning, not after reading the configuration or terminal arguments.
- Be able to be used with out a global instance, so we don't have to pass it to all sub components (it does this now.)

Nice to have:

- Scoped log levels. Like debug it would be awesome to change the log level for only one component. However this is an enhancement and nice to have.
- Time diff: https://github.com/visionmedia/debug#millisecond-diff
- Formatters https://github.com/visionmedia/debug#formatters

The great thing about not just using debugjs is that we can also log normal things that we want the user to see, and keep the syntax and other nice stuff.

Maybe we could borrow things from them, or wrap their module. But most important is that we don't rely on the configuration reader, but read **Environment Variables**.

What do you think?

**Comment by Saturate** [comment]
https://www.npmjs.com/package/debug-logger
